### PR TITLE
Extracting Let's Encrypt to a client interface

### PR DESCRIPTION
**Pull request by c-e-brumm**
This will help with dependency injection and creating mock clients.

### CODE REVIEW

File: pkg/leclient/lets_encrypt.go [hunk 1]
```diff
@@ -0,0 +1,199 @@
...
+		Client.Client, err = acme.NewClient(acme.LetsEncryptStaging)
+		return Client, err
+	}
+	Client.Client, err = acme.NewClient(acme.LetsEncryptProduction)
+	return Client, err
+}
+
+func getLetsEncryptAccountPrivateKey(kubeClient client.Client, staging bool) (privateKey crypto.Signer, err error) {
+
+	secretName := LetsEncryptProductionAccountSecretName
```
Comment (cblecker) [inline pkg/leclient/lets_encrypt.go]: Question: Do we expect an install of this operator to have both staging and prod? Is there a reason to have two different secrets in the same NS?

File: pkg/controller/certificaterequest/issue_certificate.go [hunk 2]
```diff
@@ -53,80 +54,59 @@ func (r *ReconcileCertificateRequest) IssueCertificate(reqLogger logr.Logger, cr
...
 		}
 
-		reqLogger.Info("authorization url", "authorization.URL", authorization.URL)
-
-		domain := authorization.Identifier.Value
-
-		challenge, ok := authorization.ChallengeMap[acme.ChallengeTypeDNS01]
-		if !ok {
+		domain := leClient.GetAuthorizationIndentifier()
+		leClient.SetChallengeType()
```
Comment (cblecker) [inline pkg/controller/certificaterequest/issue_certificate.go]: Uncaught error handling

File: pkg/controller/certificaterequest/issue_certificate.go [hunk 3]
```diff
@@ -53,80 +54,59 @@ func (r *ReconcileCertificateRequest) IssueCertificate(reqLogger logr.Logger, cr
...
+		if err != nil {
 			return fmt.Errorf("cloud not find DNS challenge authorization")
 		}
 
-		encodeDNS01KeyAuthorization := acme.EncodeDNS01KeyAuthorization(challenge.KeyAuthorization)
+		DNS01KeyAuthorization := leClient.GetDNS01KeyAuthorization()
```
Comment (cblecker) [inline pkg/controller/certificaterequest/issue_certificate.go]: Is it possible for this response to be `nil`?

File: pkg/controller/certificaterequest/issue_certificate.go [hunk 4]
```diff
@@ -53,80 +54,59 @@ func (r *ReconcileCertificateRequest) IssueCertificate(reqLogger logr.Logger, cr
...
 			return err
 		}
 
-		reqLogger.Info("authorization url", "authorization.URL", authorization.URL)
-
-		domain := authorization.Identifier.Value
-
-		challenge, ok := authorization.ChallengeMap[acme.ChallengeTypeDNS01]
-		if !ok {
+		domain := leClient.GetAuthorizationIndentifier()
```
Comment (cblecker) [inline pkg/controller/certificaterequest/issue_certificate.go]: Is it possible for this response to be `nil`?

File: pkg/controller/certificaterequest/issue_certificate.go [hunk 5]
```diff
@@ -53,80 +54,59 @@ func (r *ReconcileCertificateRequest) IssueCertificate(reqLogger logr.Logger, cr
...
 		certDomains = append(certDomains, domain)
-		ids = append(ids, acme.Identifier{Type: "dns", Value: domain})
 	}
 
-	letsEncryptOrder, err := letsEncryptClient.NewOrder(letsEncryptAccount, ids)
+	err = leClient.CreateOrder(cr.Spec.DnsNames)
 	if err != nil {
 		return err
 	}
+	URL, _ := leClient.GetOrderURL()
```
Comment (cblecker) [inline pkg/controller/certificaterequest/issue_certificate.go]: Why are we discarding the err here?

File: pkg/controller/certificaterequest/update_status.go [hunk 6]
```diff
@@ -42,8 +42,7 @@ func (r *ReconcileCertificateRequest) updateStatus(reqLogger logr.Logger, cr *ce
 			cr.Status.IssuerName != certificate.Issuer.CommonName ||
 			cr.Status.NotBefore != certificate.NotBefore.String() ||
 			cr.Status.NotAfter != certificate.NotAfter.String() ||
-			cr.Status.SerialNumber != certificate.SerialNumber.String() ||
-			cr.Status.Status != "Success" {
```
Comment (cblecker) [inline pkg/controller/certificaterequest/update_status.go]: This seems unrelated to the rest of the PR. Was this intended?

File: pkg/leclient/lets_encrypt.go [hunk 7]
```diff
@@ -0,0 +1,199 @@
...
+		ids = append(ids, acme.Identifier{Type: "dns", Value: domain})
+	}
+	c.Order, err = c.Client.NewOrder(c.Account, ids)
+	if err != nil {
+		return err
+	}
+	return nil
+}
+
+func (c *ACMEClient) GetAccount(kubeClient client.Client, staging bool, namespace string) (err error) {
```
[the remark this answers is not in the document]
Comment (c-e-brumm) [reply pkg/leclient/lets_encrypt.go]: This is legacy logic that I wasn't trying to remove in this PR. But you're right, there is absolutely no need for staging vs production endpoints in the operator logic. It needs to just be config data supplied to the operator (endpoint, account name, keys, etc). That will be fixed in a future PR.

File: pkg/leclient/constants.go [hunk 8]
```diff
@@ -0,0 +1,35 @@
...
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package leclient
+
+const (
```
[the remark this answers is not in the document]
Comment (c-e-brumm) [reply pkg/leclient/constants.go]: No, this is actually lifted straight from the package certificaterequest constants, where it originated, to get the pieces I needed. I forgot to curate these, good catch.

File: pkg/leclient/lets_encrypt.go (hunk 1 above)
Comment (c-e-brumm) [reply pkg/leclient/lets_encrypt.go]: No, it will never do both, which is why it needs to be streamlined out of the code. That's coming in another story/PR.

File: pkg/controller/certificaterequest/issue_certificate.go (hunk 2 above)
Comment (c-e-brumm) [reply pkg/controller/certificaterequest/issue_certificate.go]: Neither SetChallengeType or GetAuthorizationIndentifier should be able to fail but I've added error outputs and stanzas just in case in the future they can. Or in case future implementations can fail.

File: pkg/controller/certificaterequest/update_status.go (hunk 6 above)
Comment (c-e-brumm) [reply pkg/controller/certificaterequest/update_status.go]: Not related but I accidentally committed bug fix to my feature branch. This removes always catching a non "Status: Success" and setting it to "Success".

File: pkg/controller/certificaterequest/issue_certificate.go (hunk 4 above)
Comment (c-e-brumm) [reply pkg/controller/certificaterequest/issue_certificate.go]: Do not believe so. But I'll handling just in case.

File: pkg/controller/certificaterequest/issue_certificate.go (hunk 3 above)
Comment (c-e-brumm) [reply pkg/controller/certificaterequest/issue_certificate.go]: No, but check added in case.

File: pkg/controller/certificaterequest/issue_certificate.go (hunk 5 above)
Comment (c-e-brumm) [reply pkg/controller/certificaterequest/issue_certificate.go]: The left side is unchanged from before. I've fixed this and added handling.

**Comment by cblecker** [comment]
@app-sre-bot test this

**Comment by cblecker** [comment]
[test]

**Comment by cblecker** [comment]
/lgtm
/approve
/hold

Please feel free to remove hold when you are ready to merge.

File: pkg/leclient/lets_encrypt.go (hunk 7 above)
[the remark this answers is not in the document]
Comment (dofinn) [reply pkg/leclient/lets_encrypt.go]: Are tasks like this getting managed via Jira?

[the remark this answers is not in the document]
Comment (dofinn) [reply pkg/leclient/lets_encrypt.go]: https://github.com/openshift/certman-operator/issues/30

[the remark this answers is not in the document]
Comment (c-e-brumm) [reply pkg/leclient/lets_encrypt.go]: All certman-operator stories, https://jira.coreos.com/browse/SDE-109
This item specifically, https://jira.coreos.com/browse/SREP-1802
